state/machine.go

```diff
@@ -430,29 +430,33 @@ func (m *Machine) CharmProfiles() ([]string, error) {
 // SetCharmProfiles sets the names of the charm profiles used on a machine
 // in its instanceData.
 func (m *Machine) SetCharmProfiles(profiles []string) error {
-	if len(profiles) == 0 {
-		return nil
-	}
 	buildTxn := func(attempt int) ([]txn.Op, error) {
 		if attempt > 0 {
 			if err := m.Refresh(); err != nil {
 				return nil, errors.Trace(err)
 			}
 		}
+
 		// Exit early if the Machine profiles doesn't need to change.
-		mProfiles, err := m.CharmProfiles()
+		currentProfiles, err := m.CharmProfiles()
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		mProfilesSet := set.NewStrings(mProfiles...)
-		if mProfilesSet.Union(set.NewStrings(profiles...)).Size() == mProfilesSet.Size() {
+		currentProfileSet := set.NewStrings(currentProfiles...)
+		newProfileSet := set.NewStrings(profiles...)
+		if currentProfileSet.Size() == newProfileSet.Size() &&
+			currentProfileSet.Intersection(newProfileSet).Size() == currentProfileSet.Size() {
```

Comment:
**nit:** use difference and then you can check it against 0

Reply:
Difference is an exclusive left outer join. It doesn't account for the values in the right that isn't in the left set.

Reply:
This equality check exploits: If the size of the inner join set is equal to the size of both the left set and the right set, the left and right must be equal.

Reply:
If the size equality is checked (which it is), then an intersection can be dropped in favour of a difference and checking if it's nonzero.